Add KServe Fluid example on LLMs (#2819)
kserve/kserve: docs/samples/fluid/README.md

      --patch '{"data":{"kubernetes.podspec-nodeselector":"enabled"}}'
    ```
  * Enable Direct VolumeMount for PVC In KServe

    ```sh
    # edit inferenceservice-config and update enableDirectPvcVolumeMount to true
    kubectl edit configmap inferenceservice-config  -n kserve

    #   storageInitializer: |-
    #     {
    #         "image" : "kserve/storage-initializer:latest",
    #         "memoryRequest": "100Mi",
    #         "memoryLimit": "1Gi",
    #         "cpuRequest": "100m",
    #         "cpuLimit": "1",
    #         "storageSpecSecretName": "storage-config",
    #         "enableDirectPvcVolumeMount": false        # change to true
    #     }
    ```

* [Fluid](https://github.com/fluid-cloudnative/fluid/blob/master/docs/en/userguide/get_started.md)
  * Deploy Fluid

    ```sh
    # Add Fluid repository to Helm repos and keep it up-to-date
    helm repo add fluid https://fluid-cloudnative.github.io/charts
    helm repo update
    # we use the devel version here
    helm upgrade --install \
      --set webhook.reinvocationPolicy=IfNeeded \
      fluid --create-namespace --namespace=fluid-system fluid/fluid --devel
    ```

## Prepare Demo Application

> I will build a demo application using KServe to serve a LLM (Large Language Model). In this tutorial, I will use [BLOOM](https://huggingface.co/docs/transformers/model_doc/bloom).

Reviewer: ```suggestion
The demo application uses KServe to serve a LLM (Large Language Model). In this tutorial we use [BLOOM](https://huggingface.co/docs/transformers/model_doc/bloom) model as an example.
```

### Build and Push Application Image

```sh
export DOCKER_REPO='docker.io/<username>'

# build and push application image
cd docker

> docker build -f Dockerfile -t ${DOCKER_REPO}/kserve-fluid:bloom-gpu-v1 .

Reviewer: Also add a prerequisite that you need to checkout the kserve repo and run the demo under docs/samples/fluid

└── tokenizer_config.json -> ../../blobs/e7016b49fcff7e162946ec012d3c7b4db0b66d87

0 directories, 9 files
```

Upload the model to S3 bucket:
```sh
# update the path accordingly
aws s3 cp --recursive ${output_dir} s3://${bucket}/models/bloom-560m
```

### Run The Application Locally

```sh
docker run -it --rm -p 8080:8080 \
  -v $(pwd)/models:/mnt/models \
  -e MODEL_URL=/mnt/${output_dir} \
  -e MODEL_NAME=bloom \
  ${DOCKER_REPO}/kserve-fluid:bloom-gpu-v1
```

### Test The Application

```sh
curl -i -X POST -H "Content-Type: application/json" "localhost:8080/v1/models/bloom:predict" -d '{"prompt": "It was a dark and stormy night", "result_length": 50}'
```

## Demo

Let's start our demo.

### Create Demo Namespace

```sh
kubectl create ns kserve-fluid-demo

# add label to inject fluid sidecar
kubectl label namespace kserve-fluid-demo fluid.io/enable-injection=true
```

### Create S3 Credentials Secret And Service Account

```sh
# please update the s3 credentials to yours
# NOTE: for Jindo runtime, https is not supported in current version
kubectl create -f s3creds.yaml -n kserve-fluid-demo
```

### Create Dataset And Runtime

> Create Dataset and Runtime (In this tutorial I used `JindoFS` as the example, there are other runtimes like JuiceFS and Alluxio etc).

Reviewer: Nit : For documentation we usually use second person
https://developers.google.com/style/person#:~:text=In%20general%2C%20use%20second%20person,that%20your%20reader%20is%20developing.

We will use the following [machine types](https://aws.amazon.com/ec2/instance-types/m5/) for our tests:

* Data: m5.xlarge
* Workload: m5.2xlarge, m5.4xlarge

We will use the following Fluid runtimes for our tests:

* JindoFS

For the purposes of our benchmark, we will assume that nodes are readily available and that images are cached in the nodes, so we will not be factoring in node provisioning time or image pulling time.

Other assumptions:

* S3 bucket and Cluster are in the same region (eu-central-1)
* Data are preloaded in workers

Command to measure the total time:

```sh
# Total time includes: pod initialization and running + download model (storage initializer) + load model + inference + network
curl --connect-timeout 3600 --max-time 3600 -o /dev/null -s -w 'Total: %{time_total}s\n' -H "Content-Type: application/json" -H "Host: ${SERVICE_HOSTNAME}" "http://${INGRESS_HOST}:${INGRESS_PORT}/v1/models/bloom:predict" -d '{"prompt": "It was a dark and stormy night", "result_length": 50}'
# Total time: 53.037210s, status code: 200
```

> | Model Name                                                                | Model Size (Snapshot) | Machine Type | KServe + Storage Initializer                         | KServe + Fluid(JindoFS)                    |

Reviewer: @lizzzcai did you try out juicefs which seems like  a good choice as well?
https://juicefs.com/docs/community/comparison/juicefs_vs_alluxio

Author: Hi @yuzisun I didn't try it out here. One thing need to know for [juicefs](https://juicefs.com/docs/community/architecture) is you have to provide Data Storage (like S3) and Metadata Engine (like redis) as dependencies to set it up. I will give it a try as well, but like alluxio, maybe we can do it later. (I will add a link to my own [repo](https://github.com/lizzzcai/kserve-fluid-demo#performance-benchmark), and update it to this example once the data are reviewed)

|---------------------------------------------------------------------------|-----------------------|--------------|------------------------------------------------------|--------------------------------------------|

> | [ bigscience/bloom-560m ]( https://huggingface.co/bigscience/bloom-560m ) | 3.14GB                | m5.2xlarge   | total: 52.725s (download: 25.091s, load: 3.549s)     | total: 23.286s (load: 4.763s) (2 workers)  |

Reviewer: maybe break out the prediction time? I know you can do the math yourself. but might be just easier for user to see the number directly

Author: Hi @yuzisun , I will break it down to:

total: 52.725s (download: 25.091s, load: 3.549s, inference: x.xxxs, others: x.xxxs), 

inference: inference time of the first request
others: mainly include pod startup time, time spent in other steps in the application, and network latency

> | [ bigscience/bloom-7b1 ]( https://huggingface.co/bigscience/bloom-7b1 )   | 26.35GB               | m5.4xlarge   | total: 365.479s (download: 219.844s, load: 102.299s) | total: 53.037s (load: 24.137s) (3 workers) |

Reviewer: For KServe the load time means loading from empty dir to memory, for fluid it includes both getting from fluid workers and then loading to memory? just trying to understand how fluid is 2x faster(53s) than kserve(102s).

Author: For KServe, it is loading from empty dir to memory. For Fluid it is loading from volume to memory(from my understanding the reading from workers is happening concurrently). Maybe @cheyang can comment on why fluid is faster here.

Reviewer 2: @yuzisun  According to my understand, the empty dir is placed on cloud dis, so it's restricted by the single cloud disk throughput. But for the Fluid, since it has 3 workers, it can provide x3 cloud disk throughput + page cache.  I think it's the reason why fluid is faster than empty dir with larger model load.

> NOTE: `total` is the response time of sending an inference request to a serverless inference service (scale from 0), which includes pod startup, `model download`, `model load`, model inference and network time.
> TBD: add testing results for other runtimes

As can be observed, Fluid has demonstrated notable improvements in the autoscaling performance of KServe on LLM.

### Limitations

> One potential limitation of this integration is the use of an additional Fuse-sidecar, which may lead to increased resource consumption. However, in `the use case of LLMs`, where the instance size is typically very large and the pod consumes the entire instance, the resource usage of the Fuse-sidecar is expected to be negligible. Nonetheless, it is important to carefully consider the resource impact of the Fuse-sidecar in your specific use case to ensure that it remains within acceptable limits.

Reviewer: KServe also has model agent sidecar which can act as model puller, I think it is possible to integrate model agent with fluid so we do not need another `fuse-sidecar`? @lizzzcai @cheyang

Author: The fluid removes the download step in the storage initializer:

For KServe + Storage initiazlier:
- storage initializer downloads model to /mnt/models
- application loads model from /mnt/models

For KServe + Fluid:
- application loads model from /mnt/models (model is mounted by PVC, storage initializer is not needed any more)

To integrate with model-agent (which is able to download models dynamically), one possible approach is:
- model agent `dynamically` mounts object storage (for example: `model-1`) provided by user ( for example calling the `alluxio fs mount /mnt/model-1 s3://<S3_BUCKET>/<S3_DIRECTORY>/model-1`)
- model agent symlink the mount point (`/mnt/model-1`) above to `/mnt/models/model-1`, through a shared volume.

I feel like it is possible, maybe @cheyang can comment on it. However, this will need extra development works, we can put it into the next step. I am planning to restructure the folder so that future different use case can be added under `docs/samples/fluid`, the current example will be put under `docs/samples/fluid/sidecar`. @yuzisun

Reviewer: It is definitely out of scope here but would like to explore a better kserve/fluid integration as next step. I will create an issue for this.

Reviewer 2: Thank you @yuzisun . I think we can schedule a meeting for next step in May!

  * Please follow this [guideline](https://kserve.github.io/website/master/developer/developer/#deploy-kserve-from-master-branch) to deploy KServe from `master` branch
  * Enable node selector in KServe InferenceService via KNative

    ```sh

    kubectl patch configmap/config-features \
      --namespace knative-serving \
      --type merge \
      --patch '{"data":{"kubernetes.podspec-nodeselector":"enabled"}}'
    ```
  * Enable Direct VolumeMount for PVC In KServe

    ```sh
    # edit inferenceservice-config and update enableDirectPvcVolumeMount to true
    kubectl edit configmap inferenceservice-config  -n kserve

    #   storageInitializer: |-
    #     {
    #         "image" : "kserve/storage-initializer:latest",
    #         "memoryRequest": "100Mi",
    #         "memoryLimit": "1Gi",
    #         "cpuRequest": "100m",
    #         "cpuLimit": "1",
    #         "storageSpecSecretName": "storage-config",
    #         "enableDirectPvcVolumeMount": false        # change to true
    #     }
    ```

* [Fluid](https://github.com/fluid-cloudnative/fluid/blob/master/docs/en/userguide/get_started.md)
  * Deploy Fluid from `master` branch

> 

Reviewer 2: You can use the following way to install fluid latest version.

```bash
$ helm repo add fluid https://fluid-cloudnative.github.io/charts
$ helm repo update
$ helm search repo fluid --devel
NAME            CHART VERSION   APP VERSION     DESCRIPTION                               
fluid/fluid     0.9.0-alpha.7   0.9.0-ac1fb47   A Helm chart to deploy fluid in Kubernetes

$ helm install fluid fluid/fluid --devel
```

Author: Thanks @cheyang I will test it and update it accordingly.

    ```sh
    # Deploy Fluid from master branch
    git clone https://github.com/fluid-cloudnative/fluid

    # if you are using alluxio runtime
    helm upgrade --install \
        --set runtime.alluxio.enabled=true \
        --set webhook.reinvocationPolicy=IfNeeded \
        fluid --create-namespace --namespace=fluid-system fluid/charts/fluid/fluid

    # if you are using jindo runtime (this example)
    helm upgrade --install \

>         --set runtime.jindo.enabled=true \

Reviewer 2: By default, you don't have to set `runtime.jindo.enabled=true`, the runtime controller will be created according to the runtime type.
